### PR TITLE
[Draft] Implement Validation pass for train (2nd version)

### PR DESCRIPTION
**Pull request by jyoungyun**
Related issue: #12204

### CODE REVIEW

**Comment by jyoungyun** [comment]
Without `metric` argument:

```
$ ./Product/out/bin/onert_train --modelfile ../Personal/models/mnist/mnist_keras/mnist_keras_softmax2/mnist_keras.circle --load_input:raw ../Personal/models/data/fashion_mnist/train.input.1000.bin --load_expected:raw ../Personal/models/data/fashion_mnist/train.output.1000.bin --loss 1 --optimizer 1 --epoch 5 --batch_size 1000 --learning_rate 0.001 --loss_reduction_type 1
Model Expected Filename ../Personal/models/data/fashion_mnist/train.output.1000.bin
Model Input Filename ../Personal/models/data/fashion_mnist/train.input.1000.bin
Model Filename ../Personal/models/mnist/mnist_keras/mnist_keras_softmax2/mnist_keras.circle
Epoch 1/5 - time: 86.222ms/step - loss: [0] 0.0117 - val_loss: [0] 0.0114
Epoch 2/5 - time: 83.088ms/step - loss: [0] 0.0114 - val_loss: [0] 0.0111
Epoch 3/5 - time: 83.351ms/step - loss: [0] 0.0111 - val_loss: [0] 0.0109
Epoch 4/5 - time: 83.146ms/step - loss: [0] 0.0109 - val_loss: [0] 0.0106
Epoch 5/5 - time: 83.444ms/step - loss: [0] 0.0106 - val_loss: [0] 0.0103
===================================
MODEL_LOAD   takes 0.9820 ms
PREPARE      takes 14.9660 ms
EXECUTE      takes 550.5860 ms
- Epoch 1      takes 86.2220 ms
- Epoch 2      takes 83.0880 ms
- Epoch 3      takes 83.3510 ms
- Epoch 4      takes 83.1460 ms
- Epoch 5      takes 83.4440 ms
===================================
```

With `metric` argument:

```
$ ./Product/out/bin/onert_train --modelfile ../Personal/models/mnist/mnist_keras/mnist_keras_softmax2/mnist_keras.circle --load_input:raw ../Personal/models/data/fashion_mnist/train.input.1000.bin --load_expected:raw ../Personal/models/data/fashion_mnist/train.output.1000.bin --loss 1 --optimizer 1 --epoch 5 --batch_size 1000 --learning_rate 0.001 --loss_reduction_type 1 --metric 0
Model Expected Filename ../Personal/models/data/fashion_mnist/train.output.1000.bin
Model Input Filename ../Personal/models/data/fashion_mnist/train.input.1000.bin
Model Filename ../Personal/models/mnist/mnist_keras/mnist_keras_softmax2/mnist_keras.circle
Epoch 1/5 - time: 85.829ms/step - loss: [0] 0.0117 - val_loss: [0] 0.0114 - val_accuracy: [0] 0.1110
Epoch 2/5 - time: 83.381ms/step - loss: [0] 0.0114 - val_loss: [0] 0.0111 - val_accuracy: [0] 0.1110
Epoch 3/5 - time: 83.136ms/step - loss: [0] 0.0111 - val_loss: [0] 0.0109 - val_accuracy: [0] 0.1100
Epoch 4/5 - time: 83.035ms/step - loss: [0] 0.0109 - val_loss: [0] 0.0106 - val_accuracy: [0] 0.1110
Epoch 5/5 - time: 83.434ms/step - loss: [0] 0.0106 - val_loss: [0] 0.0103 - val_accuracy: [0] 0.1050
===================================
MODEL_LOAD   takes 1.0020 ms
PREPARE      takes 13.4180 ms
EXECUTE      takes 550.7950 ms
- Epoch 1      takes 85.8290 ms
- Epoch 2      takes 83.3810 ms
- Epoch 3      takes 83.1360 ms
- Epoch 4      takes 83.0350 ms
- Epoch 5      takes 83.4340 ms
===================================
```

File: runtime/onert/api/include/nnfw_experimental.h [hunk 1]
```diff
@@ -263,11 +263,33 @@ NNFW_STATUS nnfw_train_set_input(nnfw_session *session, uint32_t index, const vo
...
+ *
+ * @param[in]   session Session from inference output is to be extracted
+ * @param[in]   index   Index of output to be set (0-indexed)
+ * @param[in]   type    Type of the output
+ * @param[out]  buffer  Raw buffer for output
+ * @param[in]   length  Size of bytes of output buffer
+ *
+ * @return      @c NNFW_STATUS_NO_ERROR if successful
+ */
+NNFW_STATUS nnfw_train_set_output(nnfw_session *session, uint32_t index, NNFW_TYPE type,
```
Comment (glistening) [inline runtime/onert/api/include/nnfw_experimental.h:282]: @jyoungyun What is the difference between `nnfw_set_output` and `nnfw_train_set_output` ?

Comment (jyoungyun) [reply runtime/onert/api/include/nnfw_experimental.h:282]: The functions are the same. However, it should be invoked after `nnfw_train_prepare` because it is changes the output shape according to the batch size.

Comment (glistening) [reply runtime/onert/api/include/nnfw_experimental.h:282]: Thank you for answer. I've found we already have `nnfw_train_set_input` perhaps due to the same reason.

Comment (jyoungyun) [reply runtime/onert/api/include/nnfw_experimental.h:282]: @glistening If the same `nnfw_set_input` and `nnfw_set_output` are used in the training, the `State` class must be modified to handle both training and inference status.

And we don't need to handle batch using `nnfw_set_input_tensorinfo`. The updated batch size is already applied to the internal graph in the `nnfw_train_prepare` function. The `nnfw_train_set_output/input` function sets the user buffers to the internal graph. What I want to say in the https://github.com/Samsung/ONE/pull/12265#discussion_r1424808224 comment is that these functions should be called before `nnfw_train_prepare`, not `nnfw_prepare`. To this end, the training API is different from `nnfw_set_input/output` functions. If we change the `nnfw_set_input/output` code to check whether the `nnfw_train_prepare` has already been called, we can use the same api for training.

Anyway, introducing separate APIs for training from the inference API is our policy that we decided when we first discussed. If you think that it would be better to use the same API in training and inference, let's talk about it at our weekly meeting.

**Comment by jyoungyun** [comment]
Done!
